### ISSUE TITLE
[Regression] Incremental annotation processor fails with "Attempt to recreate a file for type"

### ISSUE THREAD

**Issue by melix** [gradle/gradle#18840]
This is related to #13767

### Expected Behavior

Incremental Java compilation should succeed.

### Current Behavior

Incremental Java compilation fails with:

> error: Unexpected error: Attempt to recreate a file for type com.example.$Controller$Definition

if using Gradle 7.3-rc-3. It compiles fine with 7.2.

### Context

I was debugging another incremental annotation processing issue and when checking what the behavior with latest Gradle was, I realized it was failing badly.

### Steps to Reproduce 

```bash
$ git clone git@github.com:melix/gradle-73-incremenal-ap-bug.git
$ cd gradle-73-incremenal-ap-bug
$ ./gradlew cJ
$ printf ' ' >> src/main/java/com/example/Application.java
$ ./gradlew cJ
```

### Your Environment

Build scan URL: https://gradle.com/s/vhw3tzhkvnebg

**Comment by melix** [comment]
I suspect the regression was introduced in https://github.com/gradle/gradle/pull/17748

**Comment by zdegner** [comment]
I was seeing this as well today with Gradle 7.3 RC3 + Java 17 + Micronaut. It seemed sporadic. The builds worked on subsequent runs.

**Comment by melix** [comment]
Subsequent runs would work because after a failure Gradle will perform full recompilation.

**Comment by melix** [comment]
I think this line might be the culprit:

https://github.com/gradle/gradle/blob/8abd7269b11795c6176be16d053e8e800955cb08/subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/deps/ClassSetAnalysis.java#L159

It unconditionally adds _all aggregated types_ as types to reprocess. It actually also maybe the reason for the performance issue I was originally investigating, which is that if you change an unrelated type (not even annotated), all aggregated types are reprocessed.

We should only reprocess types which are affected by the current changes, which is what the following loop is doing. I might need @oehme 's help on this one :)

**Comment by oehme** [comment]
I'll look into it.

**Comment by melix** [comment]
Thanks. This comment is confusing:

> types which are annotated with aggregating annotations, as aggregating processors need to see them regardless of what has changed

The APs only need to see types which would actually be affected by a change. If I change an unrelated class, then no type needs to be reprocessed. If I change a class which is involved in aggregation, then we need to reprocess all the aggregated types. So if we have:

```
@ProcessedByIsolating
class A {}

@ProcessedByIsolating
class B {}

@ProcessedByOtherIsolating
class C {}

```

which generates:

```
@ProcessedByAggregating
class A_Generated {}
@ProcessedByAggregating
class B_Generated {}
@ProcessedByOtherAggregating
class C_OtherGenerated {}
```

then if `A` is changed, we need to reprocess `A_Generated` and `B_Generated`, but not `C_OtherGenerated`.

**Comment by oehme** [comment]
It needs to see all three of them, otherwise the aggregation result would be missing an entry for `C_OtherGenerated`. In Micronaut's case, it generates a `META_INF/services` file. If we only passed `A_Generated` and `B_Generated`, then that services file would suddenly be missing `C_OtherGenerated`

**Comment by melix** [comment]
Why would it care about `C_OtherGenerated`? In my example there are 2 aggregating processors, and only one of them needs to see `A_Generated` and `B_Generated`. Currently we're reprocessing types which shouldn't have to be.

**Comment by oehme** [comment]
How would you know up-front that a change didn't introduce something new with `@ProcessedByOtherAggregating`?

**Comment by melix** [comment]
mmm, I see, yes, the problem is that we would only know during compilation itself... This is quite annoying as basically in my case incremental compilation is as slow as full recompilation :/

**Comment by oehme** [comment]
Well, it should only be reprocessing, not recompiling. But if the processing is the expensive part in your case, then yes that sucks and is unfortunately not avoidable when you need to aggregate.

**Comment by ljacomet** [comment]
@oehme @melix Based on your conversation, what should happen for 7.3? Any recommendation on how to fix this?

**Comment by oehme** [comment]
I'm already working on the fix.

**Comment by octylFractal** [comment]
Fixed by #18864
